crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+                },
+            }
+
+            self.check_leadership();
+
+            self.send_outgoing_messages();
+
+            match self.handle_decided_entries() {
+                Ok(DecidedEntriesResult::Continue) => {
+                    // nothing to do
+                }
+                Ok(DecidedEntriesResult::Stop(rocksdb_storage)) => {
+                    debug!("Stopping active metadata store node");
+                    self.kv_storage.fail_callbacks(|| {
+                        RequestError::Unavailable("stopping metadata store".into())
+                    });
+                    self.fail_join_callbacks(|| JoinClusterError::NotLeader);
+
+                    return Ok(Passive::new(
+                        rocksdb_storage,
+                        self.connection_manager,
+                        self.request_rx,
+                        self.join_cluster_rx,
+                        self.metadata_writer,
+                        self.own_member_id.storage_id,
+                    ));
+                }
+                Err(err) => {
+                    self.kv_storage.fail_callbacks(|| {
+                        RequestError::Unavailable("stopping metadata store".into())
+                    });
+                    self.fail_join_callbacks(|| JoinClusterError::NotLeader);
+                    return Err(err.into());
+                }
+            }
+        }
+    }
+
+    fn update_node_addresses(&mut self, nodes_configuration: &NodesConfiguration) {
+        for node_id in &self.cluster_config.nodes {
+            if let Ok(node_config) = nodes_configuration.find_node_by_id(PlainNodeId::from(
+                u32::try_from(*node_id).expect("node id is derived from PlainNodeId"),
+            )) {
+                self.networking
+                    .register_address(*node_id, node_config.address.clone());
+            }
+        }
+    }
+
+    fn check_leadership(&mut self) {
+        let previous_is_leader = self.is_leader;
+        self.is_leader = self
+            .omni_paxos
+            .as_ref()
+            .expect("to be present")
+            .get_current_leader()
+            .is_some_and(|(node_id, _)| node_id == self.own_member_id.node_id);
+
+        if previous_is_leader && !self.is_leader {
+            debug!(configuration_id = %self.cluster_config.configuration_id, "Lost leadership");
+
+            // we lost leadership :-( notify callers that their requests might not get committed
+            // because we don't know whether the leader will start with the same log as we have.
+            self.kv_storage
+                .fail_callbacks(|| RequestError::Unavailable("lost leadership".into()));
+            self.fail_join_callbacks(|| JoinClusterError::NotLeader);
+        } else if !previous_is_leader && self.is_leader {
+            debug!(configuration_id = %self.cluster_config.configuration_id, "Won leadership");
```

Comment:
One problem I had was that OmniPaxos does not tell you what is the lsn for a given proposal. Therefore, it is hard to track whether a proposal made it or was replaced by a different proposal (e.g. because of a leadership change and a different log suffix). That's why I am quite defensive here to avoid hanging requests.

The correlation between request and log entries is currently done by having an explicit ulid that is part of a request entry.